### ISSUE TITLE
Transition Rando

### ISSUE THREAD

**Issue by Ijwu** [Ijwu/Archipelago.HollowKnight#35]
It would be nice to have a transition randomizer to have full on room rando. We should compare and contrast ideas from current and previous HK randomizer implementations and determine the best path forward on this.

**Comment by dewiniaid** [comment]
For reference, Rando4 has these options, as I understand them

**Map Area Rando:** Transitions between areas with different maps are randomized, internal area connections are preserved.  Thus, Crossroads will look like Vanilla Crossroads, but its various exits (to Greenpath, Fog Canyon, Fungal Wastes, etc). are randomized.  This, or the next option, is probably the closest analogue to LTTP dungeon shuffle
 
**Full Area Rando:** Like the above, but based on area titles rather than maps.  The example given here is that Lake of Unn is on the Greenpath map but is a separate titled area and would be randomized separately.  I _think_ this also applies to the various Mounds (Overgrown/Ancestral/Crystallized), Watchers Spire, Path of Pain, Beast's Den, and Teachers Archives.

**Connected-Area Room Rando**: Room transitions are shuffled, but rooms from the same map area stay connected as best as possible.  So Greenpath's rooms will be reshuffled but still connected together, exiting one Greenpath room will generally put you in another one.  LTTP analogue: Doors basic.

**Room Rando:** Room transitions are shuffled.  LTTP analogue: Doors crossed.

---

TRandoplus adds the following:

**Randomize Doors:** Shuffle just doorway transitions -- i.e. anywhere you go "Up" to enter.  Notable exceptions: The inner Black Egg door that requires 3 dreamers is vanilla, as is the Abyss light house.  I'm not entirely certain of the reasons for the latter

**Randomize Drops:** Shuffle just drop transitions -- i.e. places where you fall into a void and transition.  IIRC this includes the two different ways to fall off Hallownest's Crown (top of CPeaks), the drop near Crystallized Mound into Resting Grounds, and the drops from Howling Cliffs into either King's Pass or Dirtmouth.  

---

In all of the above cases, White Palace and Path of Pain are included if they'reenabled; you may have to transit a room or two of White Palace to get somewhere.  If White Palace locations are excluded, you'll never need to transit a White Palace room.

I'm a huge fan of door rando myself, it's a nice "light" entrance rando option.

**Comment by Ijwu** [comment]
Door rando is likely to be our first implementation as all berserker has to do is match up transitions then send me the final list so I can apply the transitions.

**Comment by flibber-hk** [comment]
> I'm not entirely certain of the reasons for the latter

@dewiniaid The reason for the abyss lighthouse not being randomized in door rando is... uhh... it isn't a door - you walk right to go into the lighthouse, rather than pressing up.

**Comment by flibber-hk** [comment]
> Transitions can be decoupled, and there's some other quirks, an option for whether transitions need to match up or you can exit a room left and appear at the top of another room. If we support that latter option, some shenanigans may be required with CDash and the Knight's direction.

This shouldn't be required, actually - ItemChanger automatically fixes cdashing/diving through transitions when going the wrong way (for instance, if you dive into a down transition and get spat out a left transition, then ItemChanger stops the vanilla bug that would cause you to permanently lose control). If you play with SkillUpgrades, instead you'll dive horizontally out of the transition.

**Comment by olee** [comment]
Hey, would be interested if this is something that is still in progress somehow or if the effort is too high etc.?
I think having room rando as part of the integration would make HK undoubtedly the best Archipelago supporting game out there 👍

**Comment by BadMagic100** [comment]
Yes, it is still in progress, or more accurately, in progress again. HK was unmaintained for around a year (late 2022-late 2023) and changed hands once even after that. During that time the logic in rando4 changed substantially, we are rewriting most of the AP implementation to match. When it is done we will support random/alternative starts and transition rando.

**Comment by Nocturnal42** [comment]
So another 2 years on, how is this coming along?

**Comment by BadMagic100** [comment]
The mod supports this and the beta version of the apworld supports this. It is just waiting to be merged which may take some time, in the meantime it is playable via the pins in the discord
